### PR TITLE
[IMP] mass_mailing: support dynamic domains

### PR DESCRIPTION
**Pull request by reth-odoo**
Allow mailings to fully use front-end time-relative domains to support flows such as "send this mailing to all customers who have purchased something within the last month" which is particularly useful in the context of marketing automation.

task-4240689

### CODE REVIEW

**Comment by reth-odoo** [comment]
Security PY001 (safe_eval context should not pass in values that allow for escalation):

Only `wrap_module` modules are used (safe by construction), and no env or env-related value/reference is passed in.
Only the minimum necessary to handle domains generated by the front-end domain widget are defined.

**Review by thjo-odoo — COMMENTED** [review]
@reth-odoo 
Hello,
Nice job on this one, kinda hard to follow as really low-level, but interesting nonetheless.
From what I am understanding I didn't spot any major issues and seems robust, thanks to the tests.

Let's see what the reviewers and more experienced will think about it. I just suggested a change in a comment as it didn't make sense when I read it.

Cheers!

File: addons/web/static/src/core/domain_selector/domain_selector_operator_editor.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ export function getDomainDisplayedOperators(fieldDef) {
             ];
         case "date":
         case "datetime":
-            return ["=", "!=", ">", ">=", "<", "<=", "between", "within", "set", "not_set"];
+            return ["=", "!=", ">", ">=", "<", "<=", "between", "is_not_between", "within", "is_not_within", "set", "not_set"];
```
Comment (reth-odoo) [inline addons/web/static/src/core/domain_selector/domain_selector_operator_editor.js:40]: We don't have to add them, for the purposes of this task.

But since "not" is not automatically distributed on "complex" domains you can only get to these by typing them out entirely so they'd be low in discoverability

**Review by reth-odoo — COMMENTED** [review]
@odoo/rd-framework-js Hi there, I'm adding negative counterparts to the "between" and "within" operators in the domain widget if you'd like to give it a look.

Tried a bunch of stuff to make "negated between" appear without adding a whole operator but in the end this seemed to be the simplest. Let me know if it works for you, and whether you'd like it to show up by default or not since it's possibly niche.

**Comment by ged-odoo** [comment]
@Polymorphe57

File: addons/web/static/src/core/tree_editor/condition_tree.js [hunk 2]
```diff
@@ -901,9 +900,9 @@ export function removeWithinOperators(tree) {
         const reverse = Number.isInteger(value[0]) && value[0] > 0;
         return condition(
             path,
-            "between",
+            tree.operator === "within" ? "between" : "is_not_between",
             reverse ? Object.values(expressions) : Object.values(expressions).reverse(),
-            negate
+            negate,
```
Comment (Polymorphe57) [inline addons/web/static/src/core/tree_editor/condition_tree.js:905]: you should activate the tooling when working in those files (addons/web/tooling/enable.sh). No comma here.

**Review by Polymorphe57 — COMMENTED** [review]
The commit for web seems good. Thank you

**Comment by Polymorphe57** [comment]
If the PR is blocked for security reasons, shouldn't we already merged the  commit for web?

**Comment by reth-odoo** [comment]
I don't think it's as much security reasons as pipeline congestion ^^'

But I can split it into a separate PR sure :)

**Comment by Polymorphe57** [comment]
> I don't think it's as much security reasons as pipeline congestion ^^'
> 
> But I can split it into a separate PR sure :)

No need to do it but it would be nice to merge your work

**Comment by reth-odoo** [comment]
Alright, thanks a lot!
Removing that commit and rebasing then

**Comment by pedrobaeza** [comment]
Why not adding the feature of "relative times" to the domain editor widget? This is useful in other places I think.

**Comment by reth-odoo** [comment]
@pedrobaeza Hi
This is what "is (not) within X days" does, it's standard on date(time) fields in the domain editor. This PR just makes it work for mass_mailing/marketing_automation flows. I think it should already cover many common flows.

**Comment by pedrobaeza** [comment]
Oh, thank you very much for the reference, I didn't know about it, as I haven't dug yet enough on 18.0, which I see it includes the first one (and the not one is introduced in master). That was a usual request from customers.
